lib/test/Test/Pos/Txp/Toil/UtxoSpec.hs

```diff
@@ -456,3 +467,19 @@ txShouldFailWithPlutus res err = case res of
     other -> expectationFailure $
         "expected: Left ...: " <> show (WitnessScriptError err) <> "\n" <>
         " but got: " <> show other
+
+-- | Transaction should fail with a 'ToilInvalidWitness' specifically 'WitnessAdoptedScriptVerMismatch' error.
+txShouldFailWithInvalidWitness :: Either ToilVerFailure () -> Expectation
+txShouldFailWithInvalidWitness = \case
+    Left (ToilInvalidWitness _ _ c) -> case c of
+        WitnessAdoptedScriptVerMismatch _ _ -> pass
+        WitnessScriptVerMismatch _ _ -> expectationFailure $
+            "expected: WitnessAdoptedScriptVerMismatch\n" <>
+            " but got: WitnessScriptVerMismatch"
+        other -> expectationFailure $
+            "expected: WitnessAdoptedScriptVerMismatch\n" <>
+            " but got: " <> show other
+
+    other -> expectationFailure $
+        "expected: Left ToilInvalidWitness\n" <>
+        " but got: " <> show other
```

Comment:
It's better to use `pretty` instead of `show` (here and above).

Reply:
What would be the most correct way to do it since `expectationFailure` expects `String` and pretty produces `Text` so I need `Text -> String`

Reply:
There is `toString` in `Universum`.

Reply:
wtf when using `toString $ pretty other` I get
```
    • No instance for (Buildable (Either ToilVerFailure ()))
            arising from a use of ‘pretty’
        • In the second argument of ‘($)’, namely ‘pretty other’
          In the second argument of ‘(<>)’, namely
            ‘(toString $ pretty other)’
          In the second argument of ‘(<>)’, namely
            ‘" but got: " <> (toString $ pretty other)’
```
but I see the `Buildable` instance defined so why this happens ?

Reply:
> but I see the `Buildable` instance defined so why this happens ?

It's defined for `ToilVerFailure`, but probably not for `Either`.
You can use `toString $ either pretty show other` here. If it's `ToilVerFailure`, it will apply `pretty`, otherwise it will apply `show` to `()` which will be `"()"`.